src/server_manager/web_app/gcp_server.ts

```diff
@@ -180,7 +180,9 @@ export class GcpServer extends ShadowboxServer implements server.ManagedServer {
         const apiUrl = outlineGuestAttributes.get('apiUrl');
         try {
           const parsed = new URL(apiUrl);
-          trustCertificate(parsed.host, certSha256);
+          if (!trustCertificate(parsed.host, certSha256)) {
```

Comment:
Shouldn't we prefer GCP over manual entries? The threat is the manual entry.

Instead of trustCertificate, the call could be named `setTlsCertificateFpForService()`, which replaces a previous fingerprint. That will encode in code that a service can only have one fingerprint.

Then the manual entries should be validated in order to not overwrite FP from cloud servers

Reply:
It's not clear to me what the threat model is.  Perhaps an attacker could provide you with credentials for a cloud account containing a malicious server config.

Unfortunately, regardless of the threat model, I don't see a reasonable way to implement this kind of preference.  Once we have trusted a certificate, this is effectively irrevocable.  The network stack could keep the SSL socket open indefinitely, so that any future requests that match the `host` will be sent along that socket without triggering an SSL error that would call our event handler.  Cloud servers are necessarily added asynchronously after startup, perhaps even long after if the user is not logged in, so we cannot easily delay the loading of manual servers until after the cloud servers are processed.

If a collision is detected, the only way to prefer the _second_ server would be to disable the first server permanently and restart Electron.

Reply:
Let's tie the certificate FP to the URL instead. This way we are sure we will never send secrets to a server that can't present the right cert: https://www.electronjs.org/docs/latest/api/app#event-certificate-error

If Electron is reusing sockets as you say, I imagine Electron will call the event with the cert of the socket being used.

Reply:
I would be very surprised if that is true.  This event is "certificate error", and I expect it is only emitted during a TLS handshake, when the server sends a certificate that fails validation.  The event handler decides whether the certificate is trusted or not, which determines whether the socket is established or not.  Once the socket is established, the server does not send any further certificates, so no further "certificate errors" should be emitted for requests on that socket.

If you think it's likely that this event is actually emitted on every HTTP request that reuses this socket, we can check the behavior relatively easily.  It's also possible that Electron disables pooling entirely for these sockets, although the docs certainly don't promise that.